shakenfist/images.py

```diff
@@ -40,46 +36,83 @@ def _get_cache_path():
 
 
 class Image(object):
-    def __init__(self, url, checksum=None):
+    def __init__(self, url, checksum, size, modified, fetched, file_version):
+        self.url = url
         self.checksum = checksum
-        self.url, resolver_checksum = self._resolve(url)
-        if not self.checksum:
-            self.checksum = resolver_checksum
+        self.size = size
+        self.modified = modified
+        self.fetched = fetched
+        self.file_version = file_version
+
+        # Derive extra parameters
+        self.unique_ref = self.calc_unique_ref(self.url)
+        self.image_path = os.path.join(_get_cache_path(), self.unique_ref)
 
-        self._hash()
-        self.info = self._read_local_info()
+        self.log = LOG.withImage(self)
 
     def unique_label(self):
-        return ('image', self.hashed_image_url)
+        return ('image', self.unique_ref)
 
-    def _resolve(self, url):
+    @staticmethod
+    def _resolve(url):
         for resolver in resolvers:
             if url.startswith(resolver):
                 return resolvers[resolver].resolve(url)
         return url, None
 
-    def _hash(self):
+    @staticmethod
+    def calc_unique_ref(url):
+        """Calc unique reference for this image.
+
+        The calculated reference is used as the unique DB reference and as the
+        on-disk filename.
+        """
+
+        # TODO(andy): If we namespace downloads then this can combine namespace
+        # with the URL. The DB stores the URL allowing searches to re-use
+        # already downloaded images.
         h = hashlib.sha256()
-        h.update(self.url.encode('utf-8'))
-        self.hashed_image_url = h.hexdigest()
-        self.hashed_image_path = os.path.join(
-            _get_cache_path(), self.hashed_image_url)
-
-    def _read_local_info(self):
-        if not os.path.exists(self.hashed_image_path + '.info'):
-            LOG.withImage(self).info('No info in cache for this image')
-            return {
-                'url': self.url,
-                'hash': self.hashed_image_url,
-                'version': 0
-            }
-        else:
-            with open(self.hashed_image_path + '.info') as f:
-                return json.loads(f.read())
-
-    def _persist_info(self):
-        with open(self.hashed_image_path + '.info', 'w') as f:
-            f.write(json.dumps(self.info, indent=4, sort_keys=True))
+        h.update(url.encode('utf-8'))
+        return h.hexdigest()
+
+    @staticmethod
```

Comment:
This is interesting. In Instance and Network this would be done as a module level method called "from_url" or whatever. I wonder if this is a better approach and we should mvoe the others?

Reply:
I do suspect it should be called from_url, but whatevers.

Reply:
I agree. Consistency says I should have used snake case. Fixed.

I was hoping you would like this method to roll out to the other classes. Then we can guarantee data correctness in the object more easily.

Reply:
I would accept a patch to the other two examples to make them look like this... I do think the from_() methods should be straight after __init__ so we find them.